src/clean-core/allocator.hh

```diff
@@ -0,0 +1,211 @@
+
+    // specific interfaces which can be overriden by allocators to be more efficient, with fallback otherwise
+public:
+    /// reallocate a buffer, behaves like std::realloc
+    virtual cc::byte* realloc(void* ptr, size_t old_size, size_t new_size, size_t align = alignof(std::max_align_t));
+
+    /// allocate a buffer with specified minimum size, will span up to request_size if possible
+    [[nodiscard]] virtual cc::byte* alloc_request(size_t min_size, size_t request_size, size_t& out_received_size, size_t align = alignof(std::max_align_t));
+
+    /// reallocate a buffer with specified minimum size, will span up to request_size if possible
+    [[nodiscard]] virtual cc::byte* realloc_request(
+        void* ptr, size_t old_size, size_t new_min_size, size_t request_size, size_t& out_received_size, size_t align = alignof(std::max_align_t));
+};
+
+
+struct linear_allocator final : public allocator
+{
+public:
+    cc::byte* alloc(size_t size, size_t align = alignof(std::max_align_t)) override;
+
+    void free(void* ptr) override
+    {
+        // no-op
+        (void)ptr;
+    }
+
+    void reset() { _head = _buffer; }
+
+    size_t allocated_size() const { return _head - _buffer; }
+    size_t remaining_size() const { return _buffer_end - _head; }
+    size_t max_size() const { return _buffer_end - _buffer; }
+    float allocated_ratio() const { return allocated_size() / float(max_size()); }
+
+    linear_allocator() = default;
+    linear_allocator(cc::span<cc::byte> buffer) : _buffer(buffer.data()), _head(buffer.data()), _buffer_end(buffer.data() + buffer.size()) {}
+
+private:
+    cc::byte* _buffer = nullptr;
+    cc::byte* _head = nullptr;
+    cc::byte* _buffer_end = nullptr;
+};
+
+struct stack_allocator final : public allocator
+{
+public:
+    cc::byte* alloc(size_t size, size_t align = alignof(std::max_align_t)) override;
+
+    void free(void* ptr) override;
+
+    cc::byte* realloc(void* ptr, size_t old_size, size_t new_size, size_t align = alignof(std::max_align_t)) override;
+
+    void reset()
+    {
+        _head = _buffer;
+        _last_alloc = nullptr;
+    }
+
+    stack_allocator() = default;
+    stack_allocator(cc::span<cc::byte> buffer) : _buffer(buffer.data()), _head(buffer.data()), _buffer_end(buffer.data() + buffer.size()) {}
+
+private:
+    cc::byte* _buffer = nullptr;
+    cc::byte* _head = nullptr;
+    cc::byte* _buffer_end = nullptr;
+    cc::byte* _last_alloc = nullptr;
+};
+
+struct system_allocator_t final : public allocator
+{
+public:
+    cc::byte* alloc(size_t size, size_t align = alignof(std::max_align_t)) override;
+
+    void free(void* ptr) override;
+
+    cc::byte* realloc(void* ptr, size_t old_size, size_t new_size, size_t align = alignof(std::max_align_t)) override;
+
+    system_allocator_t() = default;
+};
+
+static cc::system_allocator_t system_allocator_instance = {};
+static cc::allocator* const system_allocator = &system_allocator_instance;
```

Comment:
`static` means that each TU gets their own instance, thus comparing system allocators of alloc_vectors from different TUs returns false

Reply:
Changed these to inline instead